src/librustc_typeck/collect.rs

```diff
@@ -916,70 +919,81 @@ fn generics_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+            }),
         }
     });
 
-    let mut types: Vec<_> = opt_self.into_iter().chain(types).collect();
+    let mut types: Vec<_> = types.into_iter().collect();
 
     // provide junk type parameter defs - the only place that
     // cares about anything but the length is instantiation,
     // and we don't do that for closures.
     if let NodeExpr(&hir::Expr { node: hir::ExprClosure(..), .. }) = node {
         // add a dummy parameter for the closure kind
-        types.push(ty::TypeParameterDef {
+        types.push(ty::GenericParamDef {
             index: type_start,
             name: Symbol::intern("<closure_kind>").as_interned_str(),
             def_id,
-            has_default: false,
-            object_lifetime_default: rl::Set1::Empty,
-            pure_wrt_drop: false,
-            synthetic: None,
+            kind: ty::GenericParamDefKind::Type(ty::TypeParamDef {
+                has_default: false,
+                object_lifetime_default: rl::Set1::Empty,
+                pure_wrt_drop: false,
+                synthetic: None,
+            }),
         });
 
         // add a dummy parameter for the closure signature
-        types.push(ty::TypeParameterDef {
+        types.push(ty::GenericParamDef {
             index: type_start + 1,
             name: Symbol::intern("<closure_signature>").as_interned_str(),
             def_id,
-            has_default: false,
-            object_lifetime_default: rl::Set1::Empty,
-            pure_wrt_drop: false,
-            synthetic: None,
+            kind: ty::GenericParamDefKind::Type(ty::TypeParamDef {
+                has_default: false,
+                object_lifetime_default: rl::Set1::Empty,
+                pure_wrt_drop: false,
+                synthetic: None,
+            }),
         });
 
         tcx.with_freevars(node_id, |fv| {
-            types.extend(fv.iter().zip(2..).map(|(_, i)| ty::TypeParameterDef {
+            types.extend(fv.iter().zip(2..).map(|(_, i)| ty::GenericParamDef {
                 index: type_start + i,
                 name: Symbol::intern("<upvar>").as_interned_str(),
                 def_id,
-                has_default: false,
-                object_lifetime_default: rl::Set1::Empty,
-                pure_wrt_drop: false,
-                synthetic: None,
+                kind: ty::GenericParamDefKind::Type(ty::TypeParamDef {
+                    has_default: false,
+                    object_lifetime_default: rl::Set1::Empty,
+                    pure_wrt_drop: false,
+                    synthetic: None,
+                }),
             }));
         });
     }
 
-    let type_param_to_index = types.iter()
-                                   .map(|param| (param.def_id, param.index))
-                                   .collect();
+    let params: Vec<_> = opt_self.into_iter()
+                                 .chain(lifetimes)
+                                 .chain(types)
+                                 .collect();
```

Comment:
Any chance this could be built up directly, without 3 components? Especially since `params[i].idx == i + parent_count` should *always* be true (and having an `assert_eq` somewhere seems useful).

Reply:
Where do you think the right place to assert this would be?